packages/web-components/src/components/ic-dialog/ic-dialog.tsx

```diff
@@ -0,0 +1,385 @@
+   */
+  @Prop() closeOnClick?: boolean = true;
+  /**
+   * If default buttons are displayed, sets the label and onclick function passed as [{label: string, onclick: string},{},{}] with empty objects for unused buttons.
+   */
+  @Prop() buttonProps?: { label: string; onclick: string }[] = [
+    { label: "Confirm", onclick: "this.icDialogConfirm()" },
+    { label: "Cancel", onclick: "this.hideDialog()" },
+    { label: "", onclick: "" },
+  ];
+
+  @State() dialogRendered: boolean = false;
+
+  @State() fadeIn: boolean = false;
+
+  /**
+   * Confirmation event for default primary button behaviour.
+   */
+  @Event() icDialogConfirmed: EventEmitter<void>;
+  icDialogConfirm = () => {
+    this.icDialogConfirmed.emit();
+  };
+
+  private sourceElement: HTMLElement;
+  private interactiveElementList: HTMLElement[];
+  private focusedElementIndex = 0;
+  private primaryButtonLabel: string;
+  private secondaryButtonLabel: string;
+  private tertiaryButtonLabel: string;
+
+  /* eslint-disable */
+
+  private primaryButtonOnclick: Function;
+  private secondaryButtonOnclick: Function;
+  private tertiaryButtonOnclick: Function;
+
+  /* eslint-enable */
+
+  /**
+   * Used to display the dialog.
+   */
+  @Method()
+  async showDialog() {
+    this.dialogRendered = true;
+    setTimeout(() => {
+      this.fadeIn = true;
+    }, 10);
+    setTimeout(() => {
+      this.setInitialFocus();
+    }, 75);
+    setTimeout(() => {
+      this.getFocusedElementIndex();
+    }, 80);
+    document.body.style.height = "100%";
+    document.body.style.overflow = "hidden";
+  }
+  /**
+   * Used to dismiss the dialog.
+   */
+  @Method()
+  async hideDialog() {
+    this.fadeIn = false;
+    setTimeout(() => {
+      this.dialogRendered = false;
+      this.sourceElement.focus();
+      document.body.style.height = "auto";
+      document.body.style.overflow = "auto";
+    }, 80);
+  }
+
+  @Listen("keydown", { target: "document" })
+  handleKeyboard(ev: KeyboardEvent): void {
+    if (this.dialogRendered) {
+      switch (ev.key) {
+        case "Tab":
+          ev.preventDefault();
+          this.focusNextInteractiveElement(ev.shiftKey);
+          break;
+        case "Escape":
+          !ev.repeat && this.hideDialog();
+          ev.stopImmediatePropagation();
+          break;
+      }
+    }
+  }
+
+  @Listen("click")
+  handleClick(ev: MouseEvent): void {
+    const dialogElement = this.el.shadowRoot.querySelector("dialog");
+    if (this.closeOnClick) {
+      if (ev.composedPath().indexOf(dialogElement) < 0) {
+        this.hideDialog();
+      }
+    }
+  }
+
+  private setInitialFocus = () => {
+    this.sourceElement = document.activeElement as HTMLElement;
+
+    let focusedElement;
+
+    if (this.el.querySelector("#focus") !== null) {
+      focusedElement = this.el.querySelector("#focus") as HTMLElement;
+    } else {
+      focusedElement = this.el.shadowRoot.querySelector(
+        "#focus"
+      ) as HTMLElement;
+    }
```

Comment:
For this to work with slotted content, doesn't there need to be a focus ID on the slotted interactive element (which I imagine would be unlikely)?

Reply:
The intention here was that users would set an id of 'focus' on whatever element they wanted to receive initial focus when using slotted content and that this would override/come before the default buttons which had 'focus' set by the component. 

Reworked this so the component now automatically adds a 'data-gets-focus' ID on whatever element should receive focus initially, including slotted content.